File: cmd/analytics/analytics.go
package analytics

import (
	"context"
	"fmt"
	"strconv"
	"strings"
	"time"

	"github.com/mailersend/mailersend-cli/internal/cmdutil"
	"github.com/mailersend/mailersend-cli/internal/output"
	"github.com/mailersend/mailersend-cli/internal/sdkclient"
	"github.com/mailersend/mailersend-go"
	"github.com/spf13/cobra"
)

var Cmd = &cobra.Command{
	Use:   "analytics",
	Short: "View email analytics",
}

func init() {
	Cmd.AddCommand(dateCmd)
	Cmd.AddCommand(countryCmd)
	Cmd.AddCommand(uaNameCmd)
	Cmd.AddCommand(uaTypeCmd)

	// date flags
	df := dateCmd.Flags()
	df.String("date-from", "", "start date as YYYY-MM-DD or unix timestamp (required)")
	df.String("date-to", "", "end date as YYYY-MM-DD or unix timestamp (required)")
	df.String("domain", "", "filter by domain name or ID")
	df.String("group-by", "", "group by: days, weeks, months, years")
	df.StringSlice("tags", nil, "filter by tags")
	df.StringSlice("event", nil, "event types to retrieve (required, min 1): queued, sent, delivered, soft_bounced, hard_bounced, opened, clicked, unsubscribed, spam_complaints")

	// country flags
	cf := countryCmd.Flags()
	cf.String("date-from", "", "start date as YYYY-MM-DD or unix timestamp (required)")
	cf.String("date-to", "", "end date as YYYY-MM-DD or unix timestamp (required)")
	cf.String("domain", "", "filter by domain name or ID")
	cf.StringSlice("tags", nil, "filter by tags")

	// ua-name flags
	uf := uaNameCmd.Flags()
	uf.String("date-from", "", "start date as YYYY-MM-DD or unix timestamp (required)")
	uf.String("date-to", "", "end date as YYYY-MM-DD or unix timestamp (required)")
	uf.String("domain", "", "filter by domain name or ID")
	uf.StringSlice("tags", nil, "filter by tags")

	// ua-type flags
	tf := uaTypeCmd.Flags()
	tf.String("date-from", "", "start date as YYYY-MM-DD or unix timestamp (required)")
	tf.String("date-to", "", "end date as YYYY-MM-DD or unix timestamp (required)")
	tf.String("domain", "", "filter by domain name or ID")
	tf.StringSlice("tags", nil, "filter by tags")
}

// --- analytics date ---

var dateCmd = &cobra.Command{
	Use:   "date",
	Short: "Get analytics grouped by date",
	RunE:  runDate,
}

func runDate(cobraCmd *cobra.Command, args []string) error {
	flags := cobraCmd.Flags()
	dateFromStr, _ := flags.GetString("date-from")
	dateToStr, _ := flags.GetString("date-to")
	events, _ := flags.GetStringSlice("event")

	if len(events) == 0 {
		return fmt.Errorf("missing required flag: --event\n\nExample:\n  mailersend-cli analytics date --date-from 2025-01-01 --date-to 2025-01-30 --event sent,delivered")
	}

	now := time.Now()
	dateFrom, dateTo, err := cmdutil.DefaultDateRange(dateFromStr, dateToStr, now)
	if err != nil {
		return err
	}

	ms, transport, err := cmdutil.NewSDKClient(cobraCmd)
	if err != nil {
		return err
	}

	domainID, _ := flags.GetString("domain")
	if domainID != "" {
		domainID, err = cmdutil.ResolveDomainSDK(ms, transport, domainID)
		if err != nil {
			return err
		}
	}
	groupBy, _ := flags.GetString("group-by")
	tags, _ := flags.GetStringSlice("tags")

	ctx := context.Background()
	result, _, err := ms.Analytics.GetActivityByDate(ctx, &mailersend.AnalyticsOptions{
		DomainID: domainID,
		DateFrom: dateFrom,
		DateTo:   dateTo,
		GroupBy:  groupBy,
		Tags:     tags,
		Event:    events,
	})
	if err != nil {
		return sdkclient.WrapError(transport, err)
	}

	if cmdutil.JSONFlag(cobraCmd) {
		return output.JSON(result)
	}

	headers := []string{"DATE"}
	for _, e := range events {
		headers = append(headers, strings.ToUpper(e))
	}

	var rows [][]string
	for _, stat := range result.Data.Stats {
		row := []string{stat.Date}
		for _, e := range events {
			row = append(row, fmt.Sprintf("%d", statValue(stat, e)))
		}
		rows = append(rows, row)
	}

	output.Table(headers, rows)
	return nil
}

// statValue extracts a named stat field from AnalyticsStats by event name.
func statValue(s mailersend.AnalyticsStats, event string) int {
	switch event {
	case "queued":
		return s.Queued
	case "sent":
		return s.Sent
	case "delivered":
		return s.Delivered
	case "soft_bounced":
		return s.SoftBounced
	case "hard_bounced":
		return s.HardBounced
	case "junk":
		return s.Junk
	case "opened":
		return s.Opened
	case "clicked":
		return s.Clicked
	case "unsubscribed":
		return s.Unsubscribed
	case "spam_complaints":
		return s.SpamComplaints
	default:
		return 0
	}
}

// --- analytics country ---

var countryCmd = &cobra.Command{
	Use:   "country",
	Short: "Get analytics grouped by country",
	RunE:  runCountry,
}

func runCountry(cobraCmd *cobra.Command, args []string) error {
	opts, err := buildOpensOptions(cobraCmd)
	if err != nil {
		return err
	}

	ctx := context.Background()
	result, _, err := opts.ms.Analytics.GetOpensByCountry(ctx, opts.options)
	if err != nil {
		return sdkclient.WrapError(opts.transport, err)
	}

	return renderOpens(cobraCmd, result, "COUNTRY", "COUNT")
}

// --- analytics ua-name ---

var uaNameCmd = &cobra.Command{
	Use:   "ua-name",
	Short: "Get analytics grouped by user agent name",
	RunE:  runUAName,
}

func runUAName(cobraCmd *cobra.Command, args []string) error {
	opts, err := buildOpensOptions(cobraCmd)
	if err != nil {
		return err
	}

	ctx := context.Background()
	result, _, err := opts.ms.Analytics.GetOpensByUserAgent(ctx, opts.options)
	if err != nil {
		return sdkclient.WrapError(opts.transport, err)
	}

	return renderOpens(cobraCmd, result, "USER AGENT", "COUNT")
}

// --- analytics ua-type ---

var uaTypeCmd = &cobra.Command{
	Use:   "ua-type",
	Short: "Get analytics grouped by user agent type",
	RunE:  runUAType,
}

func runUAType(cobraCmd *cobra.Command, args []string) error {
	opts, err := buildOpensOptions(cobraCmd)
	if err != nil {
		return err
	}

	ctx := context.Background()
	result, _, err := opts.ms.Analytics.GetOpensByReadingEnvironment(ctx, opts.options)
	if err != nil {
		return sdkclient.WrapError(opts.transport, err)
	}

	return renderOpens(cobraCmd, result, "TYPE", "COUNT")
}

// --- shared helpers for country / ua-name / ua-type ---

type opensContext struct {
	ms        *mailersend.Mailersend
	transport *sdkclient.CLITransport
	options   *mailersend.AnalyticsOptions
}

func buildOpensOptions(cobraCmd *cobra.Command) (*opensContext, error) {
	flags := cobraCmd.Flags()
	dateFromStr, _ := flags.GetString("date-from")
	dateToStr, _ := flags.GetString("date-to")

	now := time.Now()
	dateFrom, dateTo, err := cmdutil.DefaultDateRange(dateFromStr, dateToStr, now)
	if err != nil {
		return nil, err
	}

	ms, transport, err := cmdutil.NewSDKClient(cobraCmd)
	if err != nil {
		return nil, err
	}

	domainID, _ := flags.GetString("domain")
	if domainID != "" {
		domainID, err = cmdutil.ResolveDomainSDK(ms, transport, domainID)
		if err != nil {
			return nil, err
		}
	}
	tags, _ := flags.GetStringSlice("tags")

	return &opensContext{
		ms:        ms,
		transport: transport,
		options: &mailersend.AnalyticsOptions{
			DomainID: domainID,
			DateFrom: dateFrom,
			DateTo:   dateTo,
			Tags:     tags,
		},
	}, nil
}

func renderOpens(cobraCmd *cobra.Command, result *mailersend.OpensRoot, nameHeader, countHeader string) error {
	if cmdutil.JSONFlag(cobraCmd) {
		return output.JSON(result)
	}

	headers := []string{nameHeader, countHeader}
	var rows [][]string
	for _, stat := range result.Data.Stats {
		rows = append(rows, []string{stat.Name, strconv.Itoa(stat.Count)})
	}

	output.Table(headers, rows)
	return nil
}

File: src/commands/analytics.rs
use anyhow::{bail, Result};
use clap::{Args, Subcommand};
use serde_json::Value;

use crate::cli::Ctx;
use crate::output;
use crate::util::{default_date_range, jint, jstr, resolve_domain_id};

#[derive(Args)]
pub struct Cmd {
    #[command(subcommand)]
    command: Sub,
}

#[derive(Subcommand)]
enum Sub {
    /// Get analytics grouped by date
    Date {
        /// start date as YYYY-MM-DD or unix timestamp (required)
        #[arg(long, default_value = "")]
        date_from: String,
        /// end date as YYYY-MM-DD or unix timestamp (required)
        #[arg(long, default_value = "")]
        date_to: String,
        /// filter by domain name or ID
        #[arg(long, default_value = "")]
        domain: String,
        /// group by: days, weeks, months, years
        #[arg(long, default_value = "")]
        group_by: String,
        /// filter by tags
        #[arg(long, value_delimiter = ',')]
        tags: Vec<String>,
        /// event types to retrieve (required, min 1): queued, sent, delivered, soft_bounced, hard_bounced, opened, clicked, unsubscribed, spam_complaints
        #[arg(long, value_delimiter = ',')]
        event: Vec<String>,
    },
    /// Get analytics grouped by country
    Country {
        /// start date as YYYY-MM-DD or unix timestamp (required)
        #[arg(long, default_value = "")]
        date_from: String,
        /// end date as YYYY-MM-DD or unix timestamp (required)
        #[arg(long, default_value = "")]
        date_to: String,
        /// filter by domain name or ID
        #[arg(long, default_value = "")]
        domain: String,
        /// filter by tags
        #[arg(long, value_delimiter = ',')]
        tags: Vec<String>,
    },
    /// Get analytics grouped by user agent name
    UaName {
        /// start date as YYYY-MM-DD or unix timestamp (required)
        #[arg(long, default_value = "")]
        date_from: String,
        /// end date as YYYY-MM-DD or unix timestamp (required)
        #[arg(long, default_value = "")]
        date_to: String,
        /// filter by domain name or ID
        #[arg(long, default_value = "")]
        domain: String,
        /// filter by tags
        #[arg(long, value_delimiter = ',')]
        tags: Vec<String>,
    },
    /// Get analytics grouped by user agent type
    UaType {
        /// start date as YYYY-MM-DD or unix timestamp (required)
        #[arg(long, default_value = "")]
        date_from: String,
        /// end date as YYYY-MM-DD or unix timestamp (required)
        #[arg(long, default_value = "")]
        date_to: String,
        /// filter by domain name or ID
        #[arg(long, default_value = "")]
        domain: String,
        /// filter by tags
        #[arg(long, value_delimiter = ',')]
        tags: Vec<String>,
    },
}

pub fn run(ctx: &Ctx, cmd: Cmd) -> Result<()> {
    match cmd.command {
        Sub::Date {
            date_from,
            date_to,
            domain,
            group_by,
            tags,
            event,
        } => date(ctx, &date_from, &date_to, &domain, &group_by, &tags, &event),
        Sub::Country {
            date_from,
            date_to,
            domain,
            tags,
        } => opens(
            ctx,
            "/analytics/country",
            "COUNTRY",
            &date_from,
            &date_to,
            &domain,
            &tags,
        ),
        Sub::UaName {
            date_from,
            date_to,
            domain,
            tags,
        } => opens(
            ctx,
            "/analytics/ua-name",
            "USER AGENT",
            &date_from,
            &date_to,
            &domain,
            &tags,
        ),
        Sub::UaType {
            date_from,
            date_to,
            domain,
            tags,
        } => opens(
            ctx,
            "/analytics/ua-type",
            "TYPE",
            &date_from,
            &date_to,
            &domain,
            &tags,
        ),
    }
}

#[allow(clippy::too_many_arguments)]
fn date(
    ctx: &Ctx,
    date_from: &str,
    date_to: &str,
    domain: &str,
    group_by: &str,
    tags: &[String],
    events: &[String],
) -> Result<()> {
    if events.is_empty() {
        bail!("missing required flag: --event\n\nExample:\n  mailersend-cli analytics date --date-from 2025-01-01 --date-to 2025-01-30 --event sent,delivered");
    }

    let (date_from, date_to) = default_date_range(date_from, date_to)?;

    let client = ctx.client()?;

    let mut query: Vec<(&str, String)> = Vec::new();
    if !domain.is_empty() {
        let domain_id = resolve_domain_id(&client, domain)?;
        query.push(("domain_id", domain_id));
    }
    query.push(("date_from", date_from.to_string()));
    query.push(("date_to", date_to.to_string()));
    if !group_by.is_empty() {
        query.push(("group_by", group_by.to_string()));
    }
    for t in tags {
        query.push(("tags[]", t.clone()));
    }
    for e in events {
        query.push(("event[]", e.clone()));
    }

    let body = client.get_with("/analytics/date", &query)?;

    if ctx.json {
        return output::json(&body);
    }

    let mut headers: Vec<String> = vec!["DATE".to_string()];
    for e in events {
        headers.push(e.to_uppercase());
    }
    let header_refs: Vec<&str> = headers.iter().map(String::as_str).collect();

    let stats: Vec<Value> = body["data"]
        .get("stats")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default();

    let rows: Vec<Vec<String>> = stats
        .iter()
        .map(|stat| {
            let mut row = vec![jstr(stat, "date")];
            for e in events {
                row.push(jint(stat, e).to_string());
            }
            row
        })
        .collect();

    output::table(&header_refs, &rows);
    Ok(())
}

fn opens(
    ctx: &Ctx,
    path: &str,
    name_header: &str,
    date_from: &str,
    date_to: &str,
    domain: &str,
    tags: &[String],
) -> Result<()> {
    let (date_from, date_to) = default_date_range(date_from, date_to)?;

    let client = ctx.client()?;

    let mut query: Vec<(&str, String)> = Vec::new();
    if !domain.is_empty() {
        let domain_id = resolve_domain_id(&client, domain)?;
        query.push(("domain_id", domain_id));
    }
    query.push(("date_from", date_from.to_string()));
    query.push(("date_to", date_to.to_string()));
    for t in tags {
        query.push(("tags[]", t.clone()));
    }

    let body = client.get_with(path, &query)?;

    if ctx.json {
        return output::json(&body);
    }

    let stats: Vec<Value> = body["data"]
        .get("stats")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default();

    let rows: Vec<Vec<String>> = stats
        .iter()
        .map(|stat| vec![jstr(stat, "name"), jint(stat, "count").to_string()])
        .collect();

    output::table(&[name_header, "COUNT"], &rows);
    Ok(())
}
